### PR TITLE
Improve linter

### PR DESCRIPTION
**Pull request by hugoferreira**
After our discussion in #160, I've configured the linter to enforce [strict-boolean-expressions](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md) as per suggestion of @ruippeixotog . With the linter taking care that we don't do anything stupid, I'm ok in relaxing the configuration rules if you guys prefer it that way.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -43,6 +44,7 @@
     "@typescript-eslint/semi": [1, "never"],
 
     "no-console": 0,
+    "@typescript-eslint/strict-boolean-expressions": [1, {"allowString": false, "allowNumber": false, "allowNullableObject": false}],
```
Comment (ruippeixotog) [inline .eslintrc:47]: Can you please group this warning with the existing one `@typescript-eslint/semi`? All these others are actually disabled.

**Review by ruippeixotog — APPROVED** [review]
LGTM other than the comment above. My preference is to allow every boolean expression deemed safe, e.g. all the ones involving objects/arrays. I find them to be convenient and concise, but it's a question of personal preference so I'm ok either way.

File: .eslintrc [hunk 2]
```diff
@@ -17,7 +17,8 @@
   },
 
   "parserOptions": {
-    "sourceType": "module"
+    "sourceType": "module",
+    "project": "./tsconfig.json"
```
Comment (ruippeixotog) [inline .eslintrc:21]: Why was this needed? I suppose eslint assumes this to be the default.

Comment (hugoferreira) [reply .eslintrc:21]: Actually, in my computer, it fails to run without specifying the project. Let me show ya:

```text
➜  fluent-check git:(nullish-coalescing) ✗ npm run lint

> fluent-check@1.0.0 lint
> eslint --ext '.js,.ts' src test


Oops! Something went wrong! :(

ESLint: 7.21.0

Error: Error while loading rule '@typescript-eslint/strict-boolean-expressions': You have used a rule which requires parserServices to be generated. You must therefore provide a value for the "parserOptions.project" property for @typescript-eslint/parser.
Occurred while linting /Users/bytter/Development/fluent-check/src/FluentCheck.ts
    at Object.getParserServices (/Users/bytter/Development/fluent-check/node_modules/@typescript-eslint/experimental-utils/dist/eslint-utils/getParserServices.js:23:15)
    at create (/Users/bytter/Development/fluent-check/node_modules/@typescript-eslint/eslint-plugin/dist/rules/strict-boolean-expressions.js:103:37)
    at Object.create (/Users/bytter/Development/fluent-check/node_modules/@typescript-eslint/experimental-utils/dist/eslint-utils/RuleCreator.js:13:24)
    at createRuleListeners (/Users/bytter/Development/fluent-check/node_modules/eslint/lib/linter/linter.js:761:21)
    at /Users/bytter/Development/fluent-check/node_modules/eslint/lib/linter/linter.js:931:31
    at Array.forEach (<anonymous>)
    at runRules (/Users/bytter/Development/fluent-check/node_modules/eslint/lib/linter/linter.js:876:34)
    at Linter._verifyWithoutProcessors (/Users/bytter/Development/fluent-check/node_modules/eslint/lib/linter/linter.js:1175:31)
    at Linter._verifyWithConfigArray (/Users/bytter/Development/fluent-check/node_modules/eslint/lib/linter/linter.js:1273:21)
    at Linter.verify (/Users/bytter/Development/fluent-check/node_modules/eslint/lib/linter/linter.js:1228:25)
```

File: .eslintrc (hunk 1 above)
Comment (hugoferreira) [reply .eslintrc:47]: > Can you please group this warning with the existing one `@typescript-eslint/semi`? All these others are actually disabled.

Sorry, I'm not following. Are you suggesting to change it to:

`"@typescript-eslint/strict-boolean-expressions": [1, "never"]`

> You can also make it be flagged as an error instead of a warning, in the same way we disallow `==`/`!=`.

👍

**Comment by hugoferreira** [comment]
Allow me to take this opportunity to further discuss other listing rules and invite my colleagues to take a look at our coding style 😄 Things that I believe we can enable:

* Disallow unnecessary parentheses ([`no-extra-parens`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-extra-parens.md)). Currently we have 24 violations of this rule. I suggest we enable it as a warning;
* My VSCode configuration and they way the spacing works when defining objects (`[object-curly-spacing](https://eslint.org/docs/rules/object-curly-spacing)`) makes me _always_ fix a `{ a: 2, b: 3 }` to `{a: 2, b:3}`. I prefer the former, and that's the default for VSCode. If you guys prefer the later, do you know a way to change the editor to behave accordingly?
*

File: .eslintrc (hunk 2 above)
Comment (ruippeixotog) [reply .eslintrc:21]: Uh, looks like this one requires extra stuff then 👍

File: .eslintrc (hunk 1 above)
Comment (ruippeixotog) [reply .eslintrc]: That's nice, but now it's incorrectly grouped again 😅  Rules are currently grouped into level 2 (error), level 1 (warning) and level 0 (disable). You bumped this to an error, so you should put it near `eqeqeq` and `no-var`.

Also, let's be consistent on how to declare levels - we can use either numbers (0-2) or strings ("error", "warning"...), but let's not mix both.

Comment (hugoferreira) [reply .eslintrc]: Ahhhh!!! Now I see it! 😅

**Comment by ruippeixotog** [comment]
> Disallow unnecessary parentheses (no-extra-parens). Currently we have 24 violations of this rule. I suggest we enable it as a warning;

👍 

> My VSCode configuration and they way the spacing works when defining objects (object-curly-spacing) makes me always manually fix a { a: 2, b: 3 } to {a: 2, b: 3}. I prefer the former, and that's the default for VSCode. If you guys prefer the later, do you know a way to change the editor to behave accordingly?

I have a slight preference for the latter, but I'm ok either way as long as we're consistent. I thought VC Code picked eslint settings automatically just like IntelliJ, maybe there's a way to achieve that in VS Code too?

**Comment by hugoferreira** [comment]
> I thought VC Code picked eslint settings automatically just like IntelliJ, maybe there's a way to achieve that in VS Code too?

IntelliJ is really much more advanced in these things. VSCode does allow to fix linting on save, but my OCD triggers whenever I start seeing a yellow squiggle while writing code 😅 If such thing exists for VSCode, my google-fu skills are failing.

File: test/math.test.ts [hunk 3]
```diff
@@ -65,7 +65,7 @@ describe('Math properties tests', () => {
       .forall('a', fc.integer(-10, 10))
       .forall('b', fc.integer(-10, 10))
       .forall('c', fc.integer(-10, 10))
-      .then(({a, b, c}) => a + b * c === a + (b * c))
+      .then(({a, b, c}) => a + b * c === a + b * c)
       .check()
     ).to.have.property('satisfiable', true)
   })
```
Comment (hugoferreira) [inline test/math.test.ts:71]: After this linting, I now realise this test is just stupid 😅 Ok to delete it?

File: src/arbitraries/types.ts [hunk 4]
```diff
@@ -6,7 +6,7 @@ export type FluentPick<V> = {
 export type ArbitrarySize = {
   value: number
   type: 'exact' | 'estimated'
-  credibleInterval?: [number, number]
+  credibleInterval: [number, number]
```
Comment (hugoferreira) [inline src/arbitraries/types.ts:9]: I'm not sure we should force a credible interval for exact sizes... although it may help getting rid of special logic when mixing exact and estimated sizes, like in tuples (which I've marked as a ToDo).

**Review by hugoferreira — COMMENTED** [review]
I went ahead and tried to enable `no-non-null-assertion`. The modifications are well contained, and seem justified to me. What do you guys think?

**Comment by hugoferreira** [comment]
Enabling `no-unused-vars` also seemed like a no-brainer to me, and helped to catch a few stray variables.

File: test/math.test.ts (hunk 3 above)
Comment (RuiGuedes) [reply test/math.test.ts:71]: Yeah, I agree it does not make any sense now. I think it should be deleted.

**Review by RuiGuedes — APPROVED** [review]
Apart from deleting the test that you mentioned above it LGTM!

File: src/arbitraries/types.ts (hunk 4 above)
Comment (arestivo) [reply src/arbitraries/types.ts:9]: I agree with keeping credible intervals in those cases for the sake of having less special code.

**Review by arestivo — APPROVED** [review]
LGTM
